server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -100,73 +105,107 @@ class ObjectService(
+            case Some(GcsMetadata(_, lastLockedBy, expiresAt, crc32c, generation)) =>
+              val localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath)
+              for {
+                fileBody <- io.file.readAll[IO](localAbsolutePath, blockingEc, 4096).compile.to[Array]
+                calculatedCrc32c = Crc32c.calculateCrc32c(fileBody)
+                syncStatus <- if (calculatedCrc32c == crc32c) IO.pure(SyncStatus.Live) else {
+                  for {
+                    cachedGeneration <- metadataCache.get.map(_.get(req.localObjectPath))
+                    status <- cachedGeneration match {
+                      case Some(cachedGen) => if(cachedGen.generation == generation) IO.pure(SyncStatus.LocalChanged) else IO.pure(SyncStatus.RemoteChanged)
+                      case None => IO.pure(SyncStatus.Desynchronized)
+                    }
+                  } yield status
+                }
+                currentTime <- timer.clock.realTime(TimeUnit.MILLISECONDS)
+                lastLock <- expiresAt.flatTraverse[IO, WorkbenchEmail] { ea =>
+                  if (currentTime > ea.toEpochMilli)
+                    IO.pure(none[WorkbenchEmail])
+                  else
+                    IO.pure(lastLockedBy)
+                }
+              } yield MetadataResponse.EditMode(syncStatus, lastLock, generation, pair.storageLink)
+          }
+        } yield result
       }
-    } yield res
+    )
+  }
 
-  def safeDelocalize(req: SafeDelocalize): IO[Unit] =
-    for {
-      traceId <- IO(TraceId(randomUUID()))
-      storageLink <- findStorageLink(req.localObjectPath)
-      _ <- storageLink.fold[IO[Unit]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${req.localObjectPath}"))) { pair =>
-        val isSafeMode = pair._1
-        val sl = pair._2
-        if(isSafeMode)
-          IO.raiseError(SafeDelocalizeSafeModeFile(s"${req.localObjectPath} can't be delocalized since it's in safe mode"))
-        else {
-          for {
-            previousMeta <- metadataCache.get.map(_.get(req.localObjectPath))
-            fullBlobName <- IO.fromEither(getFullBlobName(req.localObjectPath, sl.cloudStorageDirectory.blobPath).leftMap(s => BadRequestException(s)))
-            gsPath = GsPath(sl.cloudStorageDirectory.bucketName, fullBlobName)
-            localAbsolutePath = config.workingDirectory.resolve(req.localObjectPath)
-            _ <- previousMeta match {
-              case Some(m) => delocalize(localAbsolutePath, gsPath, traceId, m.generation)
-              case None => delocalize(localAbsolutePath, gsPath, traceId, 0L)
-            }
-          } yield ()
-        }
+  def safeDelocalize(req: SafeDelocalize): Kleisli[IO, TraceId, Unit] = {
+    findStorageLink[Unit](
+      req.localObjectPath,
+      _ => Kleisli.liftF(IO.raiseError(SafeDelocalizeSafeModeFileError(s"${req.localObjectPath} can't be delocalized since it's in safe mode"))),
```

Comment:
what is the benefit of refactoring `safeDelocalize` like this? i'm finding it significantly harder to understand

Reply:
it's mostly to reduce duplicate code...`findStorageLink` is reused for `safeDelocalize`, `delete`, and `checkMeta`